src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
+    Tuple,
+    Unit,
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct StructDef {
     /// Fields, not including ctor
     pub fields: Vec<StructField>,
     /// ID of the constructor. This is only used for tuple- or enum-like
     /// structs.
-    pub ctor_id: Option<NodeId>,
+    pub id: NodeId,
```

Comment:
I've made a table! :)

Now:

What    | Kind  | NodeId in StructDef | NodeId in Variant
--------| ----- | ------ | ----------
Variant in Enum | Unit  | Used | N/A
Variant in Enum | Tuple | Used | N/A
Variant in Enum | Dict  | Used | N/A
Struct  | Unit  | Used | N/A
Struct  | Tuple | Used | N/A
Struct  | Dict  | Not used | N/A

After moving NodeId in Variant:

What    | Kind  | NodeId in StructDef | NodeId in Variant
--------| ----- | ------              | --------
Variant in Enum | Unit  | Some(Not used)      | Used
Variant in Enum | Tuple | Some(Not used)      | Used
Variant in Enum | Dict  | None                | Used
Struct  | Unit  | Some(Used)          | N/A
Struct  | Tuple | Some(Used)          | N/A
Struct  | Dict  | None                | N/A

Personally, I like the first variant better